include/data_sink.hpp

```diff
@@ -0,0 +1,900 @@
+    Annotated<std::string, "signal unit", Visible, Doc<"signal's physical SI unit">> signal_unit;
+    Annotated<T, "signal min", Doc<"signal physical min. (e.g. DAQ) limit">>         signal_min;
+    Annotated<T, "signal max", Doc<"signal physical max. (e.g. DAQ) limit">>         signal_max;
+
+    IN<T, std::dynamic_extent, _listener_buffer_size>                                in;
+
+    struct poller {
+        gr::circular_buffer<T>            buffer       = gr::circular_buffer<T>(_listener_buffer_size);
+        decltype(buffer.new_reader())     reader       = buffer.new_reader();
+        decltype(buffer.new_writer())     writer       = buffer.new_writer();
+        gr::circular_buffer<tag_t>        tag_buffer   = gr::circular_buffer<tag_t>(1024);
+        decltype(tag_buffer.new_reader()) tag_reader   = tag_buffer.new_reader();
+        decltype(tag_buffer.new_writer()) tag_writer   = tag_buffer.new_writer();
+        std::size_t                       samples_read = 0; // reader thread
+        std::atomic<bool>                 finished     = false;
+        std::atomic<std::size_t>          drop_count   = 0;
+
+        template<typename Handler>
+        [[nodiscard]] bool
+        process(Handler fnc) {
+            const auto available = reader.available();
+            if (available == 0) {
+                return false;
+            }
+
+            const auto read_data = reader.get(available);
+            if constexpr (requires { fnc(std::span<const T>(), std::span<const tag_t>()); }) {
+                const auto tags          = tag_reader.get();
+                const auto it            = std::find_if_not(tags.begin(), tags.end(), [until = static_cast<int64_t>(samples_read + available)](const auto &tag) { return tag.index < until; });
+                auto       relevant_tags = std::vector<tag_t>(tags.begin(), it);
+                for (auto &t : relevant_tags) {
+                    t.index -= static_cast<int64_t>(samples_read);
+                }
+                fnc(read_data, std::span<const tag_t>(relevant_tags));
+                std::ignore = tag_reader.consume(relevant_tags.size());
+            } else {
+                std::ignore = tag_reader.consume(tag_reader.available());
+                fnc(read_data);
+            }
+
+            std::ignore = reader.consume(available);
+            samples_read += available;
+            return true;
+        }
+    };
+
+    struct dataset_poller {
+        gr::circular_buffer<DataSet<T>> buffer     = gr::circular_buffer<DataSet<T>>(_listener_buffer_size);
+        decltype(buffer.new_reader())   reader     = buffer.new_reader();
+        decltype(buffer.new_writer())   writer     = buffer.new_writer();
+
+        std::atomic<bool>               finished   = false;
+        std::atomic<std::size_t>        drop_count = 0;
+
+        [[nodiscard]] bool
+        process_bulk(std::invocable<std::span<DataSet<T>>> auto fnc) {
```

Comment:
In most other cases `process_bulk(...)` and `process_one(..)` are usually XOR implementations. Which should maybe focus on one. 

Since this will be initially (mostly) needed for UI purposes, maybe @giucam and/or @ivan-cukic could comment?
@ivan-cukic

Reply:
If implementations were more complex than this, I'd say go for one, but this is trivial, so I'm fine either way - with a single function and with both process_one and process_bulk

Reply:
no opinions here either. If single function, would you prefer process_one or process_bulk?

Reply:
`process_bulk(..)` because the `process_one(...)` as defined here is not SIMD-fied, nor loop-unrolled, nor inlined.

Reply:
Ok, will do so. Here process_one() processes whole dataset though, not individual samples, not sure we can help the user with SIMD much?

Reply:
process_bulk is process() now, process_one removed